mitiq/folding_pyquil.py

```diff
@@ -1,76 +1,110 @@
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = random.sample(range(depth), reduced_depth)
+    if not (1 <= stretch <= 3):
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
+
+    if seed is not None:
+        np.random.seed(seed)
+
+    ngates = len(circuit)
+    num_to_fold = int(ngates * (stretch - 1) / 2)
+    sub_indices = np.random.choice(range(ngates), num_to_fold, replace = False)
     return fold_gates(circuit, sub_indices)
 
+def fold_gates_from_left(circuit: Program, stretch: float) -> Program:
+    """Returns a new folded circuit by applying the map G -> G G^dag G to a subset of gates corresponding to the
+    initial part of the circuit.
 
-def left_stretcher(circuit: Program, stretch: float) -> Program:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit (first part of it).
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number within 1 and 3."""
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
+
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
+
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
 
     if not ((stretch >= 1) and (stretch <= 3)):
         raise ValueError("The stretch factor must be a real number within 1 and 3.")
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = list(range(reduced_depth))
+    ngates = len(circuit)
+    num_to_fold = int(ngates * (stretch - 1) / 2)
+    sub_indices = list(range(num_to_fold))
     return fold_gates(circuit, sub_indices)
 
 
 def fold_gates(circuit: Program, sub_indices: List[int]) -> Program:
     """Applies the map G -> G G^dag G to a subset of gates of the input circuit
     determined by sub_indices."""
 
-    out = Program()
+    out = circuit.copy_everything_except_instructions()
     for j, gate in enumerate(circuit):
         out += gate
         if j in sub_indices:
             out += circuit[j : j + 1].dagger()  # trick to avoid mutating the gate
             out += gate
     return out
 
-
-def local_folding(circuit: Program, stretch: float, stretcher: Callable = left_stretcher) -> Program:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
-
+def fold_local(
+        circuit: Program,
+        stretch: float,
+        fold_method: Callable = fold_gates_from_left,
```

Comment:
```suggestion
        fold_method: Callable[[Callable, float, Tuple[Any]], Circuit] = fold_gates_from_left,
```

Reply:
^ Add type signature to `fold_method` arg.

Reply:
Ok. But isn't this the correct signature: `Callable[[Program, float, Tuple[Any]], Program]` ?

Reply:
yep that looks right @andreamari